napari/utils/notifications.py

```diff
@@ -0,0 +1,213 @@
+        severity = getattr(warning, 'severity', 'WARNING')
+        actions = getattr(warning, 'actions', ())
+        super().__init__(msg, severity, actions, type='warning')
+        self.warning = warning
+
+
+class NotificationManager:
+    """
+    A Singleton instance of notification manager.
+
+    Only one instance can be constructed; as we need notification to all flow to
+    a single location that is registered with the sys.except_hook  and
+    showwarning hook.
+
+    This can and should be used a context manager; the context manager will
+    properly rentered, and install/remove hooks and keep them in a stack to
+    restore them.
+
+    While it might seem unnecessary to make it entrant; or to make the
+    re-entrancy no-op; one need to consider that this could be used inside
+    another context manager that modify except_hook and showwarning.
+
+    Currently the original except and show warnings hooks are not called; but
+    this could be changed in the future; this poses some questions with the
+    re-entrency of the hooks themselves.
+    """
+
+    records: List[Notification]
+    _instance: 'NotificationManager' = None
+
+    def __new__(cls):
+        # singleton
+        if cls._instance is None:
+            cls._instance = object.__new__(cls)
+        else:
+            raise RuntimeError(
+                'Napari expects a single instance of NotificationManager'
+            )
+        return cls._instance
+
+    def __init__(self) -> None:
+        self.records: List[Notification] = []
+        self.exit_on_error = os.getenv('NAPARI_EXIT_ON_ERROR') in ('1', 'True')
+        self.notification_ready = self.changed = EventEmitter(
+            source=self, event_class=Notification
+        )
+        self._originals_except_hooks = []
+        self._original_showwarnings_hooks = []
+
+    def __enter__(self):
+        self.install_hooks()
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.restore_hooks()
+
+    def install_hooks(self):
+        """
+        Install a sys.excepthook and a showwarning  hook to display any message
+        in the UI, storing the previous hooks to be restored if necessary
+        """
+
+        self._originals_except_hooks.append(sys.excepthook)
+        sys.excepthook = self.receive_error
+
+        self._original_showwarnings_hooks.append(warnings.showwarning)
+        warnings.showwarning = self.receive_warning
+
+    def restore_hooks(self):
+        """
+        Remove hooks installed by `install_hooks` and restore previous hooks.
+        """
+        sys.excepthook = self._originals_except_hooks.pop()
+
+        warnings.showwarning = self._original_showwarnings_hooks.pop()
+
+    def dispatch(self, notification: Notification):
+        self.records.append(notification)
+        self.notification_ready(notification)
+
+    def receive_error(
+        self,
+        exctype: Type[Exception] = None,
+        value: Exception = None,
```

Comment:
```suggestion
        exctype: Type[BaseException] = None,
        value: BaseException = None,
```

(for signature compatibility with `sys.excepthook`